Get GPU topo information from topo cache

cmcantalupo opened the pull request:
- Fixes #2800 change request from github issues.

cmcantalupo committed:
Use PlatformTopo cache to store GPU information

- Throw in LevelZeroIOGroup constructor if ZE_AFFINITY_MASK is set
- Add generation and parsing logic for gpu info into PlatformTopo
- Use GPUInfo object only to create cache file.
- Remove GPU features from non-gpu unit tests
- Add one test for parsing GPU info from cache file
- TODO: Add back unit tests for GPU affinity

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

bgeltz commented on service/src/PlatformTopo.cpp:
```diff
@@ -525,7 +495,30 @@ namespace geopm
                                 errno ? errno : GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
             }
-
+            const GPUTopo &gtopo = gpu_topo();
```
```suggestion
            const GPUTopo &gtopo = geopm::gpu_topo();
```

bgeltz commented on service/src/PlatformTopo.cpp:
```diff
@@ -525,7 +495,30 @@ namespace geopm
                                 errno ? errno : GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
             }
-
+            const GPUTopo &gtopo = gpu_topo();
```
Somewhere in here there should be a throw to see if ZE_AFFINITY_MASK is set.  It's unfortunate that it's a L0 specific env var.  Not sure if NVML has a similar feature.

cmcantalupo replied:
I just put the check in the constructor for the LevelZeroGPUTopo.

bgeltz commented on service/src/PlatformTopo.cpp:
```diff
@@ -525,7 +495,30 @@ namespace geopm
+                cache_stream.open(tmp_path, std::ios_base::app);
+                std::vector<int> gpu_domains = {
+                    GEOPM_DOMAIN_GPU,
+                    GEOPM_DOMAIN_PACKAGE_INTEGRATED_GPU,
```
```suggestion
```

bgeltz commented on service/src/PlatformTopo.cpp:
```diff
@@ -630,6 +623,31 @@ namespace geopm
         }
+        return numa_map;
+    }
+
```
```suggestion
```

bgeltz commented on service/src/PlatformTopo.cpp:
```diff
@@ -115,23 +116,20 @@ namespace geopm
-        parse_lscpu_numa(lscpu_map, m_numa_map);
+        m_numa_map = parse_lscpu_numa(lscpu_map);
+        m_gpu_info[GEOPM_DOMAIN_GPU] = parse_lscpu_gpu(lscpu_map, GEOPM_DOMAIN_GPU);
+        m_gpu_info[GEOPM_DOMAIN_PACKAGE_INTEGRATED_GPU] = parse_lscpu_gpu(lscpu_map, GEOPM_DOMAIN_PACKAGE_INTEGRATED_GPU);
```
```suggestion
```

bgeltz commented on service/src/PlatformTopo.cpp:
```diff
@@ -115,23 +116,20 @@ namespace geopm
-        , m_gpu_topo(gpu_topo)
+        , m_gpu_info {
+              {GEOPM_DOMAIN_GPU, {}},
+              {GEOPM_DOMAIN_PACKAGE_INTEGRATED_GPU, {}},
```
```suggestion
```

bgeltz commented on service/src/PlatformTopo.cpp:
```diff
@@ -630,6 +623,31 @@ namespace geopm
+                gpu_map.push_back({});
+                auto cpu_set_it = gpu_map.end() - 1;
+                for (const auto &cpu_str : string_split(lscpu_it->second, ",")) {
+                    cpu_set_it->insert(stoi(cpu_str));
```
```suggestion
                    cpu_set_it->insert(std::stoi(cpu_str));
```

bgeltz commented on service/src/PlatformTopo.cpp:
```diff
@@ -630,6 +623,31 @@ namespace geopm
+            else {
+                gpu_map.push_back({});
+                auto cpu_set_it = gpu_map.end() - 1;
+                for (const auto &cpu_str : string_split(lscpu_it->second, ",")) {
```
```suggestion
                for (const auto &cpu_str : geopm::string_split(lscpu_it->second, ",")) {
```

cmcantalupo committed:
Isloate service vs. client topo caches

- The service, running as root, will always and forever
  perfer to write and parse the cache under
  /run/geopm-service/geopm-topo-cache.
- When any user is calling in here to create the cache,
  always and forever perfer to write and parse the
  cache under /tmp/geopm-topo-cache-<UID>.

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Move ZE_AFFINITY_MASK exception

- The old location is too late to be of use.
- The constructor of the IOGroup creates a
  LevelZero device pool, which will try to
  call zeInit().  This all occurs before
  the IOGroup's constructor fires.

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

cmcantalupo commented on service/src/PlatformTopo.cpp:
```diff
@@ -431,10 +431,10 @@ namespace geopm
     void PlatformTopoImp::create_cache(void)
     {
-        try {
+        if (getuid() == 0) {
```
geteuid()?

cmcantalupo commented on service/src/PlatformTopo.cpp:
```diff
@@ -699,14 +699,14 @@ namespace geopm
-                result = geopm::read_file(M_CACHE_FILE_NAME);
-            }
-            catch (const geopm::Exception &ex) {
+            if (getuid() == 0) {
```
geteuid()?

cmcantalupo committed:
Add unit test for GPU data generation for topo cache

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

bgeltz commented on service/src/PlatformTopo.cpp:
```diff
@@ -485,6 +456,7 @@ namespace geopm
         }
 
         if (is_file_ok == false) {
+
```
```suggestion
```

bgeltz committed:
Apply suggestions from code review

- Remove whitespace
